crates/polars-core/src/hashing/partition.rs

```diff
@@ -9,87 +9,89 @@ pub trait AsU64 {
     fn as_u64(self) -> u64;
 }
 
+// Identity gives a poorly distributed hash. So we spread values around by a simple wrapping
+// multiplication by a 'random' odd number.
+const RANDOM_ODD: u64 = 0x55fbfd6bfc5458e9;
+
 impl AsU64 for u8 {
     #[inline]
     fn as_u64(self) -> u64 {
-        self as u64
+        (self as u64).wrapping_mul(RANDOM_ODD)
     }
 }
 
 impl AsU64 for u16 {
     #[inline]
     fn as_u64(self) -> u64 {
-        self as u64
+        (self as u64).wrapping_mul(RANDOM_ODD)
     }
 }
 
 impl AsU64 for u32 {
     #[inline]
     fn as_u64(self) -> u64 {
-        self as u64
+        (self as u64).wrapping_mul(RANDOM_ODD)
     }
 }
 
 impl AsU64 for u64 {
     #[inline]
     fn as_u64(self) -> u64 {
-        self
+        self.wrapping_mul(RANDOM_ODD)
     }
 }
 
 impl AsU64 for i8 {
     #[inline]
     fn as_u64(self) -> u64 {
-        self as u64
+        (self as u64).wrapping_mul(RANDOM_ODD)
     }
 }
 
 impl AsU64 for i16 {
     #[inline]
     fn as_u64(self) -> u64 {
-        self as u64
+        (self as u64).wrapping_mul(RANDOM_ODD)
     }
 }
 
 impl AsU64 for i32 {
     #[inline]
     fn as_u64(self) -> u64 {
         let asu32: u32 = unsafe { std::mem::transmute(self) };
-        asu32 as u64
+        (asu32 as u64).wrapping_mul(RANDOM_ODD)
     }
 }
 
 impl AsU64 for i64 {
     #[inline]
     fn as_u64(self) -> u64 {
-        unsafe { std::mem::transmute(self) }
+        (unsafe { std::mem::transmute::<_, u64>(self) }).wrapping_mul(RANDOM_ODD)
```

Comment:
Can we get rid of these transmutes? They're completely unnecessary, can just do `self as u64`.

Reply:
Yeap..